### PR TITLE
General cleanup of package:file

### PR DESCRIPTION
**Pull request by tvolkert**
* Fix/update analysis options
* Fix some linter errors
* Run dartfmt
* Fix some Dart 2 strong-mode type checks
* Remove usage of "part" and "part of" in the forwarding, memory, and local backends.
* Fix AppVeyor and Travis scripts to work only for Dart 2
* Comment out test that's causing problems with newer SDKs

### CODE REVIEW

File: test/common_tests.dart [hunk 1]
```diff
@@ -2232,23 +2235,27 @@ void runCommonTests(
             expect(await f.readAsString(), 'Hello world\n');
           });
 
+          // TODO(tvolkert): Fix and re-enable: http://dartbug.com/29554
```
Comment (goderbauer) [inline test/common_tests.dart]: Maybe mark this test as skipped instead?

File: pubspec.yaml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 name: file
-version: 4.0.0
+version: 4.0.1
```
[the remark this answers is not in the document]
Comment (tvolkert) [reply pubspec.yaml]: No, that was a previous bump (see the sdk section below).  This is just the first time we've informed Travis and AppVeyor 😄

File: test/common_tests.dart [hunk 3]
```diff
@@ -2232,23 +2235,27 @@ void runCommonTests(
...
+          /*
           test('ignoresDataWrittenAfterClose', () async {
             sink.write('Before close');
             await closeSink();
             sink.write('After close');
             expect(await f.readAsString(), 'Before close');
           });
```
[the remark this answers is not in the document]
Comment (tvolkert) [reply test/common_tests.dart]: We hijack the `test()` command in this file to enable us to call into this from various file system tests, so `skip` isn't supported here.

File: test/common_tests.dart (hunk 1 above)
Comment (tvolkert) [reply test/common_tests.dart]: We hijack the test() command in this file to enable us to call into this from various file system tests, so skip isn't supported here.

File: lib/src/backends/memory/memory_file_system.dart [hunk 4]
```diff
@@ -43,31 +28,45 @@ typedef _Node _SegmentVisitor(
...
-class MemoryFileSystem extends FileSystem {
-  _RootNode _root;
+/// any `dart:io` APIs; it merely uses the library's enum values and interfaces.
+/// As such, it is suitable for use in the browser.
+abstract class MemoryFileSystem implements FileSystem {
+  /// Creates a new `MemoryFileSystem`.
+  ///
+  /// The file system will be empty, and the current directory will be the
+  /// root directory.
+  factory MemoryFileSystem() => new _MemoryFileSystem();
```
[the remark this answers is not in the document]
Comment (tvolkert) [reply lib/src/backends/memory/memory_file_system.dart]: I was Googling for that yesterday and couldn't find the syntax - thanks!

**Pull request merged**